qualtran/bloqs/arithmetic/controlled_addition_test.py

```diff
@@ -0,0 +1,71 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import cirq
+import pytest
+
+import qualtran.testing as qlt_testing
+from qualtran import QUInt
+from qualtran.bloqs.arithmetic.controlled_addition import ControlledAdd
+from qualtran.cirq_interop.bit_tools import iter_bits
+from qualtran.cirq_interop.testing import assert_circuit_inp_out_cirqsim
+from qualtran.resource_counting.generalizers import ignore_split_join
+
+
+@pytest.mark.parametrize('a', [1, 2])
+@pytest.mark.parametrize('b', [1, 2, 3])
+@pytest.mark.parametrize('num_bits_a', [2, 3])
+@pytest.mark.parametrize('num_bits_b', [5])
+@pytest.mark.parametrize('controlled_on', [0, 1])
+@pytest.mark.parametrize('control', [0, 1])
+def test_controlled_addition(a, b, num_bits_a, num_bits_b, controlled_on, control):
+    num_anc = num_bits_b - 1
+    gate = ControlledAdd(QUInt(num_bits_a), QUInt(num_bits_b), controlled=controlled_on)
+    qubits = cirq.LineQubit.range(num_bits_a + num_bits_b + 1)
+    op = gate.on_registers(ctrl=qubits[0], a=qubits[1 : num_bits_a + 1], b=qubits[num_bits_a + 1 :])
+    greedy_mm = cirq.GreedyQubitManager(prefix="_a", maximize_reuse=True)
+    context = cirq.DecompositionContext(greedy_mm)
+    circuit = cirq.Circuit(cirq.decompose_once(op, context=context))
+    circuit0 = cirq.Circuit(op)
+    ancillas = sorted(circuit.all_qubits())[-num_anc:]
+    initial_state = [0] * (num_bits_a + num_bits_b + num_anc + 1)
+    initial_state[0] = control
+    initial_state[1 : num_bits_a + 1] = list(iter_bits(a, num_bits_a))
+    initial_state[num_bits_a + 1 : num_bits_a + num_bits_b + 1] = list(iter_bits(b, num_bits_b))
+    final_state = [0] * (num_bits_a + num_bits_b + num_anc + 1)
+    final_state[0] = control
+    final_state[1 : num_bits_a + 1] = list(iter_bits(a, num_bits_a))
+    if control == controlled_on:
+        final_state[num_bits_a + 1 : num_bits_a + num_bits_b + 1] = list(
+            iter_bits(a + b, num_bits_b)
+        )
+    else:
+        final_state[num_bits_a + 1 : num_bits_a + num_bits_b + 1] = list(iter_bits(b, num_bits_b))
+    assert_circuit_inp_out_cirqsim(circuit, qubits + ancillas, initial_state, final_state)
+    assert_circuit_inp_out_cirqsim(
+        circuit0, qubits, initial_state[:-num_anc], final_state[:-num_anc]
+    )
+    # Test diagrams
+    expected_wire_symbols = (
+        ("In(ctrl)",) + ("In(x)",) * num_bits_a + ("In(y)/Out(x+y)",) * num_bits_b
+    )
+    assert cirq.circuit_diagram_info(gate).wire_symbols == expected_wire_symbols
+
+
+@pytest.mark.parametrize("n", [*range(3, 10)])
+def test_addition_gate_counts_controlled(n: int):
+    add = ControlledAdd(QUInt(n), controlled=1)
+    qlt_testing.assert_valid_bloq_decomposition(add)
+    assert add.t_complexity() == add.decompose_bloq().t_complexity()
+    assert add.bloq_counts() == add.decompose_bloq().bloq_counts(generalizer=ignore_split_join)
```

Comment:
Another thing to check is the formula for the expected value (addition is n - 1 toffolis, controlled addition is n toffolis).

Reply:
I added a test for the Toffoli count, although its not exactly n toffolis.

Reply:
where's the test?

Reply:
assert add.t_complexity().t == t_count line 74

Reply:
I thought controlled addition was supposed to be `n` toffolis (per above). The formula you're comparing against is 8*(n-1) + 4. What are these coefficients? If we assume 4 T gates per toffoli, this is a factor of two different

Reply:
Controlled addition takes twice as many T / Toffoli's as a non-controlled addition. See https://arxiv.org/pdf/1709.06648 for reference. 

The confusion is being caused because of the way the code is structured (decomposition and `build_composite_graph` is fine, `_t_complexity_` should be refactored to have `4 * n_and` instead of `8 * n_toffoli`) - 1 $\text{And} / \text{And}^{\dagger}$ pair is hiding inside the `MultiControlPauli` bloq which corresponds to the rightmost $\text{And}  - CNOT - \text{And}^{\dagger}$ in Fig (a)

<img width="1038" alt="image" src="https://github.com/quantumlib/Qualtran/assets/7863287/f789a55e-77db-4f3c-925b-c9afef8c2b7b">

Reply:
^^ my bad, it's ~ + 1 more toffolis per bit, I misremembered.

Reply:
Ok, I fixed it to be based on the number of and pairs. I think what I have now should be correct.